applications/GeoMechanicsApplication/custom_utilities/element_utilities.hpp

```diff
@@ -249,132 +223,55 @@ typedef std::size_t IndexType;
-                                            const BoundedMatrix<double,TDim*TNumNodes, TDim*TNumNodes> &UBlockMatrix)
-    {
-        for (unsigned int i = 0; i < TNumNodes; ++i) {
-            const unsigned int Global_i = i * (TDim + 1);
-            const unsigned int Local_i  = i * TDim;
-
-            for (unsigned int j = 0; j < TNumNodes; ++j) {
-                const unsigned int Global_j = j * (TDim + 1);
-                const unsigned int Local_j  = j * TDim;
-
-                for (unsigned int idim = 0; idim < TDim; ++idim) {
-                    for (unsigned int jdim = 0; jdim < TDim; ++jdim) {
-                        rLeftHandSideMatrix(Global_i+idim, Global_j+jdim) += UBlockMatrix(Local_i+idim, Local_j+jdim);
-                    }
-                }
-            }
-        }
-    }
-
-//----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-    static inline void AssembleUBlockMatrix(Matrix &rLeftHandSideMatrix,
-                                            const Matrix &UBlockMatrix,
-                                            unsigned int TNumNodes,
-                                            unsigned int TDim)
-    {
-        for (unsigned int i = 0; i < TNumNodes; ++i) {
-            const unsigned int Global_i = i * (TDim + 1);
-            const unsigned int Local_i  = i * TDim;
-
-            for (unsigned int j = 0; j < TNumNodes; ++j) {
-                const unsigned int Global_j = j * (TDim + 1);
-                const unsigned int Local_j  = j * TDim;
-
-                for (unsigned int idim = 0; idim < TDim; ++idim) {
-                    for (unsigned int jdim = 0; jdim < TDim; ++jdim) {
-                        rLeftHandSideMatrix(Global_i+idim, Global_j+jdim) += UBlockMatrix(Local_i+idim, Local_j+jdim);
-                    }
-                }
-            }
-        }
+    template <typename MatrixType1, typename MatrixType2>
+    static inline void AssembleUUBlockMatrix(MatrixType1& rLeftHandSideMatrix, const MatrixType2& rUBlockMatrix)
+    {
+        constexpr auto row_offset    = std::size_t{0};
+        constexpr auto column_offset = row_offset;
+        AddMatrixAtPosition(rUBlockMatrix, rLeftHandSideMatrix, row_offset, column_offset);
     }
 
-//----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-    template< unsigned int TDim, unsigned int TNumNodes >
+    template <unsigned int TDim, unsigned int TNumNodes>
     static inline void AssembleUPBlockMatrix(Matrix& rLeftHandSideMatrix,
-                                             const BoundedMatrix<double,TDim*TNumNodes,TNumNodes>& UPBlockMatrix)
+                                             const BoundedMatrix<double, TDim * TNumNodes, TNumNodes>& rUPBlockMatrix)
```

Comment:
For the matrix dimension, I wonder if it is better to refer to a number of displacement D.O.F. and a number of water_pressure D.O.F. With TNumNodes here as second size, I suspect that we need another one for the diff order element.

Reply:
Here, you've touched upon a much bigger issue. In a nutshell, what I have changed is that the interfaces of the assemble functions are now similar to the one of `AssembleUUBlockMatrix`. That is, the template types have been changed, and all functions have been updated. Since we know where the submatrix/subvector needs to be added, we can either hard-code it (when no offset is required) or calculate the offset(s) from the dimensions of the given matrices/vectors.

Again, this cleans up a lot of code. Thanks for bringing this up.